feature/dynamodb/attributevalue/decode.go

```diff
@@ -0,0 +1,711 @@
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeList(avList []types.AttributeValue, v reflect.Value) error {
+	switch v.Kind() {
+	case reflect.Slice:
+		// Make room for the slice elements if needed
+		if v.IsNil() || v.Cap() < len(avList) {
+			// What about if ignoring nil/empty values?
+			v.Set(reflect.MakeSlice(v.Type(), 0, len(avList)))
+		}
+	case reflect.Array:
+		// Limited to capacity of existing array.
+	case reflect.Interface:
+		s := make([]interface{}, len(avList))
+		for i, av := range avList {
+			if err := d.decode(av, reflect.ValueOf(&s[i]).Elem(), tag{}); err != nil {
+				return err
+			}
+		}
+		v.Set(reflect.ValueOf(s))
+		return nil
+	default:
+		return &UnmarshalTypeError{Value: "list", Type: v.Type()}
+	}
+
+	// If v is not a slice, array
+	for i := 0; i < v.Cap() && i < len(avList); i++ {
+		v.SetLen(i + 1)
+		if err := d.decode(avList[i], v.Index(i), tag{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeMap(avMap map[string]types.AttributeValue, v reflect.Value) error {
+	switch v.Kind() {
+	case reflect.Map:
+		t := v.Type()
+		if t.Key().Kind() != reflect.String {
+			return &UnmarshalTypeError{Value: "map string key", Type: t.Key()}
+		}
+		if v.IsNil() {
+			v.Set(reflect.MakeMap(t))
+		}
+	case reflect.Struct:
+	case reflect.Interface:
+		v.Set(reflect.MakeMap(stringInterfaceMapType))
+		v = v.Elem()
+	default:
+		return &UnmarshalTypeError{Value: "map", Type: v.Type()}
+	}
+
+	if v.Kind() == reflect.Map {
+		for k, av := range avMap {
+			key := reflect.New(v.Type().Key()).Elem()
+			key.SetString(k)
+			elem := reflect.New(v.Type().Elem()).Elem()
+			if err := d.decode(av, elem, tag{}); err != nil {
+				return err
+			}
+			v.SetMapIndex(key, elem)
+		}
+	} else if v.Kind() == reflect.Struct {
+		fields := unionStructFields(v.Type(), structFieldOptions{
+			TagKey: d.options.TagKey,
+		})
+		for k, av := range avMap {
+			if f, ok := fieldByName(fields, k); ok {
+				fv := fieldByIndex(v, f.Index, func(v *reflect.Value) bool {
+					v.Set(reflect.New(v.Type().Elem()))
+					return true // to continue the loop.
+				})
+				if err := d.decode(av, fv, f.tag); err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeNull(v reflect.Value) error {
+	if v.IsValid() && v.CanSet() {
+		v.Set(reflect.Zero(v.Type()))
+	}
+
+	return nil
+}
+
+func (d *Decoder) decodeString(s string, v reflect.Value, fieldTag tag) error {
+	if fieldTag.AsString {
+		return d.decodeNumber(s, v, fieldTag)
+	}
+
+	// To maintain backwards compatibility with ConvertFrom family of methods which
+	// converted strings to time.Time structs
+	if v.Type().ConvertibleTo(timeType) {
+		t, err := time.Parse(time.RFC3339, s)
+		if err != nil {
+			return err
+		}
+		v.Set(reflect.ValueOf(t).Convert(v.Type()))
+		return nil
+	}
+
+	switch v.Kind() {
+	case reflect.String:
+		v.SetString(s)
+	case reflect.Interface:
```

Comment:
V1 had another case here for `reflect.Slice` for base64 encoded binary data. This was to maintain compatibility from the ConvertFrom family. should that be included here? I noticed you kept the ConvertFrom behavior of times above, so curious on this particular removal.

It's also interesting that we didn't support this in V1 on the encoder path though, making this feature asymmetric.

Reply:
Ref v1 change: https://github.com/aws/aws-sdk-go/commit/b608d29f7aa2b2d05e6bfd1c4ed160557c02b8f7#diff-ce8c08c3b104c532c9c755ce29719c4fbb9b666570edadbf4937ff94fd8210c9

Reply:
In discussion decided to leave out this change for now, since it was a backwards compatibility with the deprecated Convert utilities within v1.